In [1]:
import pandas as pd
import sqlite3

# 9.1 Reading data from SQL databases

So far we've only talked about reading data from CSV files. That's a pretty common way to store data, but there are many others! Pandas can read from HTML, JSON, SQL, Excel (!!!), HDF5, Stata, and a few other things. In this chapter we'll talk about reading data from SQL databases.

You can read data from a SQL database using the `pd.read_sql` function. `read_sql` will automatically convert SQL column names to DataFrame column names.

`read_sql` takes 2 arguments: a `SELECT` statement, and a database connection object. This is great because it means you can read from *any* kind of SQL database -- it doesn't matter if it's MySQL, SQLite, PostgreSQL, or something else.

This example reads from a SQLite database, but any other database would work the same way.

In [3]:
con = sqlite3.connect("../data/weather_2012.sqlite")
df = pd.read_sql("SELECT * from weather_2012 where temp < 2.0", con)
df

id            date_time  temp
0     1  2012-01-01 00:00:00  -1.8
1     2  2012-01-01 01:00:00  -1.8
2     3  2012-01-01 02:00:00  -1.8
3     4  2012-01-01 03:00:00  -1.5
4     5  2012-01-01 04:00:00  -1.5
5     6  2012-01-01 05:00:00  -1.4
6     7  2012-01-01 06:00:00  -1.5
7     8  2012-01-01 07:00:00  -1.4
8     9  2012-01-01 08:00:00  -1.4
9    10  2012-01-01 09:00:00  -1.3
10   11  2012-01-01 10:00:00  -1.0
11   12  2012-01-01 11:00:00  -0.5
12   13  2012-01-01 12:00:00  -0.2
13   14  2012-01-01 13:00:00   0.2
14   15  2012-01-01 14:00:00   0.8
15   16  2012-01-01 15:00:00   1.8
16   33  2012-01-02 08:00:00   1.9
17   34  2012-01-02 09:00:00   1.8
18   35  2012-01-02 10:00:00   1.5
19   37  2012-01-02 12:00:00   1.7
20   38  2012-01-02 13:00:00   1.1
21   39  2012-01-02 14:00:00   1.1
22   40  2012-01-02 15:00:00   0.0
23   41  2012-01-02 16:00:00  -0.7
24   42  2012-01-02 17:00:00  -2.1
25   43  2012-01-02 18:00:00  -4.1
26   44  2012-01-02 19:00:00  -4.8
27   45  2012-01-02 20:00:00  -5.6
28   46  2012-01-02 21:00:00  -5.8
29   47  2012-01-02 22:00:00  -7.0
..  ...                  ...   ...
53   71  2012-01-03 22:00:00 -17.1
54   72  2012-01-03 23:00:00 -17.3
55   73  2012-01-04 00:00:00 -17.5
56   74  2012-01-04 01:00:00 -17.9
57   75  2012-01-04 02:00:00 -18.1
58   76  2012-01-04 03:00:00 -18.5
59   77  2012-01-04 04:00:00 -18.5
60   78  2012-01-04 05:00:00 -18.6
61   79  2012-01-04 06:00:00 -18.5
62   80  2012-01-04 07:00:00 -18.2
63   81  2012-01-04 08:00:00 -17.8
64   82  2012-01-04 09:00:00 -16.8
65   83  2012-01-04 10:00:00 -15.2
66   84  2012-01-04 11:00:00 -14.2
67   85  2012-01-04 12:00:00 -13.7
68   86  2012-01-04 13:00:00 -12.4
69   87  2012-01-04 14:00:00 -11.3
70   88  2012-01-04 15:00:00 -10.2
71   89  2012-01-04 16:00:00  -9.4
72   90  2012-01-04 17:00:00  -8.9
73   91  2012-01-04 18:00:00  -8.9
74   92  2012-01-04 19:00:00  -8.4
75   93  2012-01-04 20:00:00  -7.8
76   94  2012-01-04 21:00:00  -7.6
77   95  2012-01-04 22:00:00  -9.5
78   96  2012-01-04 23:00:00  -9.6
79   97  2012-01-05 00:00:00  -8.8
80   98  2012-01-05 01:00:00  -7.5
81   99  2012-01-05 02:00:00  -5.4
82  100  2012-01-05 03:00:00  -5.0

[83 rows x 3 columns]

`read_sql` doesn't automatically set the primary key (`id`) to be the index of the dataframe. You can make it do that by adding an `index_col` argument to `read_sql`. 

If you've used `read_csv` a lot, you may have seen that it has an `index_col` argument as well. This one behaves the same way.

In [4]:
df = pd.read_sql("SELECT * from weather_2012 LIMIT 3", con, index_col='id')
df

date_time  temp
id                           
1   2012-01-01 00:00:00  -1.8
2   2012-01-01 01:00:00  -1.8
3   2012-01-01 02:00:00  -1.8

If you want your dataframe to be indexed by more than one column, you can give a list of columns to `index_col`:

In [5]:
df = pd.read_sql("SELECT * from weather_2012 LIMIT 3", con, 
                 index_col=['id', 'date_time'])
df

temp
id date_time                
1  2012-01-01 00:00:00  -1.8
2  2012-01-01 01:00:00  -1.8
3  2012-01-01 02:00:00  -1.8

# 9.2 Writing to a SQLite database

Pandas has a `write_frame` function which creates a database table from a dataframe. Right now this only works for SQLite databases. Let's use it to move our 2012 weather data into SQL.

You'll notice that this function is in `pd.io.sql`. There are a ton of useful functions for reading and writing various kind of data in `pd.io`, and it's worth spending some time exploring them. ([see the documentation!](http://pandas.pydata.org/pandas-docs/stable/io.html))

In [6]:
weather_df = pd.read_csv('../data/weather_2012.csv')
con = sqlite3.connect("../data/test_db.sqlite")
con.execute("DROP TABLE IF EXISTS weather_2012")
weather_df.to_sql("weather_2012", con)

/usr/local/lib/python2.7/dist-packages/pandas/core/generic.py:1362: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  chunksize=chunksize, dtype=dtype)


We can now read from the `weather_2012` table in  `test_db.sqlite`, and we see that we get the same data back:

In [8]:
con = sqlite3.connect("../data/test_db.sqlite")
df = pd.read_sql("SELECT * from weather_2012 LIMIT 3", con)
df

index            Date/Time  Temp (C)  Dew Point Temp (C)  Rel Hum (%)  \
0      0  2012-01-01 00:00:00      -1.8                -3.9           86   
1      1  2012-01-01 01:00:00      -1.8                -3.7           87   
2      2  2012-01-01 02:00:00      -1.8                -3.4           89   

   Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)               Weather  
0                4              8.0           101.24                   Fog  
1                4              8.0           101.24                   Fog  
2                7              4.0           101.26  Freezing Drizzle,Fog

The nice thing about having your data in a database is that you can do arbitrary SQL queries. This is cool especially if you're more familiar with SQL. Here's an example of sorting by the Weather column:

In [9]:
con = sqlite3.connect("../data/test_db.sqlite")
df = pd.read_sql("SELECT * from weather_2012 ORDER BY Weather LIMIT 3", con)
df

index            Date/Time  Temp (C)  Dew Point Temp (C)  Rel Hum (%)  \
0     67  2012-01-03 19:00:00     -16.9               -24.8           50   
1    114  2012-01-05 18:00:00      -7.1               -14.4           56   
2    115  2012-01-05 19:00:00      -9.2               -15.4           61   

   Wind Spd (km/h)  Visibility (km)  Stn Press (kPa) Weather  
0               24             25.0           101.74   Clear  
1               11             25.0           100.71   Clear  
2                7             25.0           100.80   Clear

If you have a PostgreSQL database or MySQL database, reading from it works exactly the same way as reading from a SQLite database. You create a connection using `psycopg2.connect()` or `MySQLdb.connect()`, and then use

`pd.read_sql("SELECT whatever from your_table", con)`

# 9.3 Connecting to other kinds of database

To connect to a MySQL database:

*Note: For these to work, you will need a working MySQL / PostgreSQL database, with the correct localhost, database name, etc.*

To connect to a PostgreSQL database:

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  